---
title: "S. pennellii 3Prime DGE simulations"
author: "Julin Maloof"
date: "January 23, 2016"
output:
  revealjs::revealjs_presentation:
    reveal_options:
      width: 1200
      height: 800
---

```{r setup, include=FALSE}
library(knitr)
opts_chunk$set(cache=TRUE)
opts_chunk$set(warnings=FALSE)
opts_chunk$set(messages=FALSE)
```

# Introduction

## Overall Idea

I want to determine how well various aligners perform when mapping S. pennellii RNAseq reads to the Heinz genome.  The basic method will be to simulate reads from fasta files, map, and then compare results to reality.

This script is for 3' DGE

# Preliminaries

## load libraries and helper scripts
```{r,cache=FALSE}
library(plyr)
library(polyester)
library(Biostrings)
library(ggplot2)
library(Rsamtools)
library(rtracklayer)
library(Rsubread)
library(R.utils)
if (system("hostname",intern=TRUE) == "whitney") 
  setwd("~/2015/rnaseq_simulations") else 
  setwd("~/Documents/Lab Notebook support/2015/rnaseq_simulations")
source("Rnaseq_sim_helper_functions.R")
library(data.table)
library(msa)
library(snowfall)
```

## Generate 3'UTR transcripts (S. pen)
Use Mike's Script

Spen CDS fasta can be obtained from [SGN](ftp://ftp.solgenomics.net/genomes/Solanum_pennellii/annotations/)
slyc fasta is also from [SGN](ftp://ftp.sgn.cornell.edu/genomes/Solanum_lycopersicum/annotation/ITAG2.4_release/)

Only needs to be done once
```{r, eval=FALSE}
system(paste("~/git/extract-utr/extract-utr.pl",
             "--threeprime",
             "--cds_fa_file=/Users/jmaloof/Sequences/ref_genomes/S.pen/Spenn-v2-cds-annot.fa",
             "--gff_file=/Users/jmaloof/Sequences/ref_genomes/S.pen/spenn_v2.0_gene_models_annot.gff",
             "--genome_fa_file=/Users/jmaloof/Sequences/ref_genomes/S.pen/Spenn.fasta",
             "--output_fa_file=/Users/jmaloof/Sequences/ref_genomes/S.pen/Spenn-v2-3primeutr-250-250.fa",
             "--utr_length=250",
             "--gene_length=250"
))
```

## lycopersiucm 3' UTR transcripts
```{r, eval=FALSE}
system(paste("~/git/extract-utr/extract-utr.pl",
             "--threeprime",
             "--cds_fa_file=/Users/jmaloof/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4_cds.fasta",
             "--gff_file=/Users/jmaloof/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4_gene_models.gff3",
             "--genome_fa_file=/Users/jmaloof/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/S_lycopersicum_chromosomes.2.50.fa",
             "--output_fa_file=/Users/jmaloof/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4-threeprimeutr-250-250.fa",
             "--utr_length=250",
             "--gene_length=250"
))
```

## Modify ITAG gff
For read counting to work appropriately the Heinz GFF file needs to be modified to match the expected reads (i.e. +/- 250 from stop codon)
```{r mod_gff,eval=FALSE}
library(rtracklayer)
heinz.gff <- readGFF("~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4_gene_models.gff3")
three.prime.df <- rbindlist(lapply(heinz.gff$Name[heinz.gff$type=="gene"], function(gene) {
  tmp <- as.data.frame(heinz.gff[grep(gene,heinz.gff$ID,fixed=T),])
  if (nlevels(factor(tmp$strand))>1) stop(paste("cannot detemine strand for gene",gene))
  if (all(tmp$strand=="+")) {
    stop.pos <- max(tmp$end[tmp$type=="CDS"])
  } else if (all(tmp$strand=="-")) {
    stop.pos <- min(tmp$start[tmp$type=="CDS"])
  } 
  result <- rbindlist(list(tmp[tmp$type=="gene",][1,],
                           tmp[tmp$type=="mRNA",][1,],
                           tmp[tmp$type=="exon",][1,]) )
  result$start <- stop.pos - 250 #this seems backwards for - strand, 
  #relative to the data.frame but it is required for converting to GRanges
  result$end <- stop.pos + 250
  as.data.frame(result)
}))
three.prime.GR <- makeGRangesFromDataFrame(three.prime.df,keep.extra.columns = TRUE)
export(three.prime.GR,"~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4_three_prime_250_250.gff3","gff3")
```


# Simluate DGE reads (simple version)

## Load Spen and Slyc 3' UTR  

```{r load_Spen, eval=TRUE}
pen.transcripts <-  readDNAStringSet("~/Sequences/ref_genomes/S.pen/Spenn-v2-3primeutr-250-250.fa")
head(pen.transcripts)
names(pen.transcripts[1:10])
lyc.transcripts <- readDNAStringSet("~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4-threeprimeutr-250-250.fa")
head(lyc.transcripts)
names(lyc.transcripts)[1:10]
```

## Reformat names
Get rid of description and add Slyc reference
```{r reformat_gene_names, eval=TRUE}
orthologs <- read.delim("Ortho_v2.txt",header=FALSE) #From Tony Bolger
head(orthologs)
names(orthologs) <- c("PEN","LYC")
orthologs$PEN <- sub("t","",orthologs$PEN)
orthologs$PEN.LYC <- paste(orthologs$PEN,orthologs$LYC,sep="_")
names(pen.transcripts) <- substr(names(pen.transcripts),1,16)
names(pen.transcripts) <- ifelse(!is.na(match(names(pen.transcripts),orthologs$PEN)),
                                 orthologs$PEN.LYC[match(names(pen.transcripts),orthologs$PEN)],
                                 names(pen.transcripts))
pen.transcripts

names(lyc.transcripts) <- substr(names(lyc.transcripts),1,18)
names(lyc.transcripts) <- ifelse(!is.na(match(names(lyc.transcripts),orthologs$LYC)),
                                 orthologs$PEN.LYC[match(names(lyc.transcripts),orthologs$LYC)],
                                 names(lyc.transcripts))
pen.gene.length <- as.integer(nchar(pen.transcripts))
names(pen.gene.length) <- names(pen.transcripts)
lyc.gene.length <- as.integer(nchar(lyc.transcripts))
names(lyc.gene.length) <- names(lyc.transcripts)
```

## Limit ourselves to where orthology is clear:
```{r reduce_data, eval=TRUE}
pen.transcripts.small <- pen.transcripts[nchar(names(pen.transcripts))>16]
writeXStringSet(pen.transcripts.small,file="~/Sequences/ref_genomes/S.pen/Spenn-v2-3primeutr-250-250-orthologs-only.fa")
lyc.transcripts.small <- lyc.transcripts[nchar(names(lyc.transcripts)) > 18]
writeXStringSet(lyc.transcripts.small,file="~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4-threeprimeutr-250-250-orthlogs-only.fa")
```

## Simulate PEN RNAseq reads
```{r simulate_pen_1, eval=FALSE}
#15 minutes for 5M reads
system.time(
  simulate_experiment(fasta="~/Sequences/ref_genomes/S.pen/Spenn-v2-3primeutr-250-250-orthologs-only.fa",
                      outdir="spen_threeprime_sim_1",
                      num_reps=1,
                      readlen=45,
                      #line below gives ~ 5M reads, scaled by transcript length
                      reads_per_transcript = round(width(pen.transcripts.small) * 
                                                     5e6/ sum(width(pen.transcripts.small) )),
                      paired=FALSE,
                      error_model="illumina5",
                      bias="rnaf"
  )
)
system("sed 's!/!;!' spen_threeprime_sim_1/sample_01.fasta > spen_threeprime_sim_1/sample_01_renamed.fasta") #this allows original gene_names to be kept
system("rm spen_threeprime_sim_1/sample_01.fasta")
system("gzip spen_threeprime_sim_1/sample_01_renamed.fasta")
```

## Simulate LYC RNAseq reads
```{r simulate_lyc_1, eval=FALSE}
system.time(
  simulate_experiment(fasta="~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4-threeprimeutr-250-250-orthlogs-only.fa",
                      outdir="slyc_threeprime_sim_1",
                      num_reps=1,
                      readlen=45,
                      #line below gives ~ 5M reads, scaled by transcript length
                      reads_per_transcript = round(width(lyc.transcripts.small) * 
                                                     5e6/ sum(width(pen.transcripts.small) )),
                      paired=FALSE,
                      error_model="illumina5",
                      bias="rnaf"
  )
)
system("sed 's!/!;!' slyc_threeprime_sim_1/sample_01.fasta > slyc_threeprime_sim_1/sample_01_renamed.fasta") #this allows original gene_names to be kept
system("rm slyc_threeprime_sim_1/sample_01.fasta")
system("gzip slyc_threeprime_sim_1/sample_01_renamed.fasta")
```

## calculate known counts
Use external python script
```{r known_counts}
pen.known.counts <- read.table(pipe("./CountGene.py spen_threeprime_sim_1/sample_01_renamed.fasta.gz"),col.names=c("id","count"),stringsAsFactors = FALSE)
pen.known.counts$lyc.id <- substr(pen.known.counts$id,18,33)
head(pen.known.counts)

lyc.known.counts <- read.table(pipe("./CountGene.py slyc_threeprime_sim_1/sample_01_renamed.fasta.gz"),col.names=c("id","count"),stringsAsFactors = FALSE)
lyc.known.counts$lyc.id <- substr(lyc.known.counts$id,18,33)
head(lyc.known.counts)
```

# Map with star

## Create STAR genome file
Only needs to be done once  
This is a shell command  
Takes about 25 minutes  
```{r create_genome_file, eval=FALSE, engine='bash'}
gffread -T ITAG2.4_gene_models.gff3 -o ITAG2.4_gene_models.gtf
STAR --runThreadN 6 --runMode genomeGenerate --genomeDir ~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/STAR_genome --genomeFastaFiles ~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/S_lycopersicum_chromosomes.2.50.fa --sjdbGTFfile ~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4_gene_models.gtf
```

## Create parameters files
```{r STAR_params_1, eval=FALSE, engine='bash'}
cat > STAR.params.whitney.1
genomeDir /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/jmaloof/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/STAR_genome
outSAMtype BAM SortedByCoordinate
outSAMunmapped Within

cat > STAR.params.royce.1
genomeDir /Users/jmaloof/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/STAR_genome
outSAMtype BAM SortedByCoordinate
outSAMunmapped Within
```

## Run it!
```{r RunStar_1, eval=FALSE}
RunStar(fasta="sample_01_renamed.fasta.gz",prefix = "STAR_1_map_2_genome",dir="spen_threeprime_sim_1",param.file = "../STAR.params.whitney.1")
RunStar(fasta="sample_01_renamed.fasta.gz",prefix = "STAR_1_map_2_genome",dir="slyc_threeprime_sim_1",param.file = "../STAR.params.whitney.1")
```
__note__ I needed to use STAR2.5 for this to work.

# Evaluate results STAR map to genome

## Load in mapped counts
```{r get_mapped_counts }
#using featureCounts from Rsubread
pen.mapped.counts <- GetMappedCounts(prefix = "STAR_1_map_2_genome",
                                     dir = "spen_threeprime_sim_1",
                                     type="STAR-featureCounts",
                                    gff="~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4_three_prime_250_250.gff3")

lyc.mapped.counts <- GetMappedCounts(prefix = "STAR_1_map_2_genome",
                                     dir = "slyc_threeprime_sim_1",
                                     type="STAR-featureCounts",
                                     gff="~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4_three_prime_250_250.gff3")
```

## Comparison (_S.pen_)
```{r compare_spen1}
pen.count.comparison.1 <- CompareCounts(pen.known.counts, pen.mapped.counts, title = "S. pen.  STAR mapped.  Default parameters.  FeatureCounts counting",chrom.separate = FALSE)
```

## Comparison (_S. lyc_)
```{r compare_slyc1}
lyc.count.comparison.1 <- CompareCounts(lyc.known.counts, lyc.mapped.counts, title = "S. lyc.  STAR mapped.  Default parameters FeatureCounts counting",chrom.separate = FALSE)
```


# Map with STAR to Heinz 3' DGE sequences (instead of whole genome)

## Create "genome" file
Only needs to be done once  
This is a shell command  
Takes about 5 minutes  
```{r create_3prime_genome_file, eval=FALSE, engine='bash'}
STAR --runThreadN 2 --runMode genomeGenerate --genomeDir ~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/STAR_3prime --genomeFastaFiles ~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4-threeprimeutr-250-250.fa
```

## Create parameters files
```{r STAR_params_3prime_1, eval=FALSE, engine='bash'}
cat > STAR.params.whitney.3prime
genomeDir /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/jmaloof/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/STAR_3prime
outSAMtype BAM SortedByCoordinate
outSAMunmapped Within

cat > STAR.params.royce.3prime
genomeDir /Users/jmaloof/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/STAR_3prime
outSAMtype BAM SortedByCoordinate
outSAMunmapped Within
```

## Run it!
```{r RunStar_1_3prime, eval=FALSE}
RunStar(fasta="sample_01_renamed.fasta.gz",
        prefix = "STAR_1_map_2_3prime",
        dir="spen_threeprime_sim_1",
        param.file = "../STAR.params.royce.3prime",
        n=2)
RunStar(fasta="sample_01_renamed.fasta.gz",
        prefix = "STAR_1_map_2_3prime",
        dir="slyc_threeprime_sim_1",
        param.file = "../STAR.params.royce.3prime",
        n=2)
```

## Load in mapped counts
```{r get_mapped_counts_3prime }
#using featureCounts from Rsubread
pen.mapped.counts.to3prime <- GetMappedCounts(bam = "STAR_1_map_2_3primeAligned.sortedByCoord.out.bam",
                                     dir = "spen_threeprime_sim_1",
                                     type="transcripts",
                                    gff="~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4_three_prime_250_250.gff3")

lyc.mapped.counts.to3prime <- GetMappedCounts(bam = "STAR_1_map_2_3primeAligned.sortedByCoord.out.bam",
                                     dir = "slyc_threeprime_sim_1",
                                     type="transcripts",
                                     gff="~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4_three_prime_250_250.gff3")
```

## Comparison (_S.pen_)
```{r compare_spen1_3DGE}
pen.count.comparison.1.to3prime <- CompareCounts(pen.known.counts, pen.mapped.counts.to3prime, title = "S. pen. STAR mapped to 3prime.  Default parameters.  FeatureCounts counting",chrom.separate = FALSE)
```

## Comparison (_S. lyc_)
```{r compare_slyc1_3DGE}
lyc.count.comparison.1.to3prime <- CompareCounts(lyc.known.counts, lyc.mapped.counts.to3prime, title = "S. lyc. STAR mapped to 3prime.  Default parameters FeatureCounts counting",chrom.separate = FALSE)
```

## eliminate overlapping genes

Next steps: some comparative genomic alignments for the genes that aren't behaving well.  (ideas: BLAT produces PSL files and IGV can read those.  Or use STAR.  Or use BLAST and the blast2sam).

## BLAT
```{r run_BLAT, engine='bash', eval=FALSE}
blat ~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/S_lycopersicum_chromosomes.2.50.fa -q=rna -t=dna -minMatch=4 ~/Sequences/ref_genomes/S.pen/Spenn-v2-3primeutr-250-250-orthologs-only.fa -maxIntron=20000 pen_3prime_to_lycGENME.psl
```

## find overlaps
```{r, eval=TRUE}
lyc.3prime.gr <- import("~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4_three_prime_250_250.gff3",feature.type="mRNA")
lyc.lyc.overlaps <- findOverlaps(lyc.3prime.gr,lyc.3prime.gr) #all overlaps including self on self
lyc.lyc.overlaps <- lyc.lyc.overlaps[queryHits(lyc.lyc.overlaps)!=subjectHits(lyc.lyc.overlaps),] #limit to non-self overlaps
lyc.lyc.overlaps <- substr(unique(mcols(lyc.3prime.gr)[queryHits(lyc.lyc.overlaps),"Name"]),1,16)

pen.3prime.gr <- PSL2Granges(file="pen_3prime_to_lycGENME.psl")
pen.pen.overlaps <- findOverlaps(pen.3prime.gr,pen.3prime.gr) #all overlaps including self on self

#need additional step to limit to non-self overlaps because BLAT returns multiple hits in the same region.  Make sure that the overlap is with the same gene
pen.pen.overlaps <- pen.pen.overlaps[names(pen.3prime.gr)[queryHits(pen.pen.overlaps)]!=names(pen.3prime.gr)[subjectHits(pen.pen.overlaps)]] 

pen.pen.overlaps <- unique(mcols(pen.3prime.gr)[queryHits(pen.pen.overlaps),"lyc.ID"]) 
```

## flag overlaps
```{r, eval=TRUE}
pen.count.comparison.1$flag <- "none"
pen.count.comparison.1 <- within(pen.count.comparison.1,{

  flag[pen.count.comparison.1$lyc.id %in% lyc.lyc.overlaps] <- "lyc.lyc.overlap"
  flag[pen.count.comparison.1$lyc.id %in% pen.pen.overlaps] <- "pen.pen.overlap"
}
)
with(pen.count.comparison.1,cor(count.known,count.mapped))
with(pen.count.comparison.1[pen.count.comparison.1$flag=="none",],cor(count.known,count.mapped))
```

not much improvement.

# kallisto to 3'

## build index
only needs to be done once:
```{r, eval=FALSE, engine='bash'}
kallisto index -k 19 -i ITAG2.4._3primeCDSUTR.pen.orthlogsonly.19.kai ~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4-threeprimeutr-250-250-orthlogs-only.fa
```

## For kallisto fasta needs to be converted to fastq
Download [converter](https://code.google.com/archive/p/fasta-to-fastq/downloads)
```{r, engine='bash',eval=FALSE}
gzcat  spen_threeprime_sim_1/sample_01_renamed.fasta.gz | perl ~/bin/fasta_to_fastq.pl  - | gzip > spen_threeprime_sim_1/sample_01_renamed.fastq.gz
gzcat  slyc_threeprime_sim_1/sample_01_renamed.fasta.gz | perl ~/bin/fasta_to_fastq.pl  - | gzip > slyc_threeprime_sim_1/sample_01_renamed.fastq.gz
```

## run kallisto
```{r run_kallisto_19, eval=TRUE}
RunKallisto(fasta = "spen_threeprime_sim_1/sample_01_renamed.fastq.gz", index = "ITAG2.4._3primeCDSUTR.pen.orthlogsonly.19.kai", dir = "spen_threeprime_sim_1/kallisto19" )
```

## Load in mapped counts
```{r get_mapped_counts_kallisto19 }
pen.mapped.counts <- GetMappedCounts(dir = "spen_threeprime_sim_1/kallisto19", type="kallisto-transcripts")
```

## _S. pennellii_
Compare known to mapped counts  
```{r compare_spen1_kallisto19, echo=FALSE, results='hide', fig.width=8, fig.height=5}
pen.count.comparison.1 <- CompareCounts(pen.known.counts, pen.mapped.counts, chrom.separate=FALSE, title = "S. pen. kallisto mapped to 3_prime_500bp 19nt kmer.  Kallisto counting")  
```

## Remove overlapping genes and plot
Genes with more than one _S.lyc_ or more than one _S.pen_ annotation overlap are removed.
```{r remove_bad_kallisto_19, echo=FALSE}
pen.count.comparison.1$flag <- "none"
pen.count.comparison.1 <- within(pen.count.comparison.1,{
                                 flag[pen.count.comparison.1$lyc.id %in% lyc.lyc.overlaps] <- "lyc.lyc.overlap"
                                 flag[pen.count.comparison.1$lyc.id %in% pen.pen.overlaps] <- "pen.pen.overlap"
                                 }
                                 )
```

```{r compare_spen1_kallisto19_bad_removed, echo=FALSE, results='hide', fig.width=8, fig.height=5}
CompareCounts(pen.count.comparison.1[pen.count.comparison.1$flag=="none",], chrom.separate = FALSE, title = "S. pen. kallisto mapped to 3_prime_500bp  19nt kmer.  Kallisto counting.  Overlapping genes removed.")  
```

# kallisto to full transcripts

## run kallisto
```{r run_kallisto_19_full, eval=TRUE}
RunKallisto(fasta = "spen_threeprime_sim_1/sample_01_renamed.fastq.gz", index = "ITAG2.4._cds.pen.orthlogsonly.19.kai", dir = "spen_threeprime_sim_1/kallisto19_cDNA" )
```

## Load in mapped counts
```{r get_mapped_counts_kallisto19_full }
pen.mapped.counts <- GetMappedCounts(dir = "spen_threeprime_sim_1/kallisto19_cDNA", type="kallisto-transcripts")
```

## _S. pennellii_
Compare known to mapped counts  
```{r compare_spen1_kallisto19_full, echo=FALSE, results='hide', fig.width=8, fig.height=5}
pen.count.comparison.1 <- CompareCounts(pen.known.counts, pen.mapped.counts, chrom.separate=FALSE, title = "S. pen. kallisto mapped to cDNA. 21nt kmer.  Kallisto counting")  
```

## Remove overlapping genes and plot
Genes with more than one _S.lyc_ or more than one _S.pen_ annotation overlap are removed.
```{r remove_bad_kallisto_19_full, echo=FALSE}
pen.count.comparison.1$flag <- "none"
pen.count.comparison.1 <- within(pen.count.comparison.1,{
                                 flag[pen.count.comparison.1$lyc.id %in% lyc.lyc.overlaps] <- "lyc.lyc.overlap"
                                 flag[pen.count.comparison.1$lyc.id %in% pen.pen.overlaps] <- "pen.pen.overlap"
                                 }
                                 )
```

## fix me I Am removing the wrong genes
```{r compare_spen1_kallisto21_bad_removed, echo=FALSE, results='hide', fig.width=8, fig.height=5}
CompareCounts(pen.count.comparison.1[pen.count.comparison.1$flag=="none",], chrom.separate = FALSE, title = "S. pen. kallisto mapped to cDNA.  21nt kmer.  Kallisto counting.  Overlapping genes removed.")  
```

# kallisto to Brad assemblies

## Background

Brad used Abyss to make assemblies from the 3' reads.  I will BLAT these against the genome to assign best hits and then map to them.  I will also try an alternative approach, blat against Heinz cDNA.

Download files from coloma.plb.ucdavis.edu/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/bradtownsley/Assemblies_Sneo_Spen

## Load contigs and take a look

```{r penn_contigs}
penn_contigs <- readDNAStringSet("Assemblies_Sneo_Spen/pen_Contigs_Singlets.longer_than_100.fa")
length(penn_contigs)
hist(log10(nchar(penn_contigs)),main="Length of S. pen 3' contigs")
```

## Blat Against Heinz Genomic

psl output
```{r blat_penn_contigs_genome, engine='bash'}
blat ~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/S_lycopersicum_chromosomes.2.50.fa Assemblies_Sneo_Spen/pen_Contigs_Singlets.longer_than_100.fa -q=rna -t=dna -minMatch=4 -maxIntron=20000 -dots=1000 Assemblies_Sneo_Spen/pen_3prime_contigs_to_lyc_genome.psl
```

also BLAST-like output
```{r blat_penn_contigs_genome_blast_output, engine='bash'}
blat ~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/S_lycopersicum_chromosomes.2.50.fa Assemblies_Sneo_Spen/pen_Contigs_Singlets.longer_than_100.fa -q=rna -t=dna -minMatch=4 -maxIntron=20000 -out=blast Assemblies_Sneo_Spen/pen_3prime_contigs_to_lyc_genome.blat.out
```

## Blat against Heinz cDNA

```{r blat_penn_contigs_cDNA, engine='bash'}
blat ~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4_cds.fasta Assemblies_Sneo_Spen/pen_Contigs_Singlets.longer_than_100.fa -q=rna -t=dna -minMatch=4 -maxIntron=20000 -dots=1000 Assemblies_Sneo_Spen/pen_3prime_contigs_to_lyc_cDNA.psl
```

BLAST-like output

```{r blat_penn_contigs_cDNA_blast_output, engine='bash'}
blat ~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4_cds.fasta Assemblies_Sneo_Spen/pen_Contigs_Singlets.longer_than_100.fa -q=rna -t=dna -minMatch=4 -maxIntron=20000 -out=blast Assemblies_Sneo_Spen/pen_3prime_contigs_to_lyc_cDNA.blat.out
```

## Blat Stats

```{r load blat_penn_genome, echo=FALSE, results='hide'}
blat_penn.results.genome <- read.delim("Assemblies_Sneo_Spen/pen_3prime_contigs_to_lyc_genome.psl",skip=5,stringsAsFactors = FALSE,header=FALSE)
blat_penn.results.genome_header <- read.delim("Assemblies_Sneo_Spen/pen_3prime_contigs_to_lyc_genome.psl",skip=2,nrows=2,stringsAsFactors = FALSE,header=FALSE)
names(blat_penn.results.genome) <- make.names(paste(trimws(blat_penn.results.genome_header[1,]),trimws(blat_penn.results.genome_header[2,]),sep=""))
head(blat_penn.results.genome)
summary(blat_penn.results.genome)
```

```{r load blat_penn_cDNA, echo=FALSE, results='hide'}
blat_penn.results.cDNA <- read.delim("Assemblies_Sneo_Spen/pen_3prime_contigs_to_lyc_cDNA.psl",skip=5,stringsAsFactors = FALSE,header=FALSE)
blat_penn.results.cDNA_header <- read.delim("Assemblies_Sneo_Spen/pen_3prime_contigs_to_lyc_cDNA.psl",skip=2,nrows=2,stringsAsFactors = FALSE,header=FALSE)
names(blat_penn.results.cDNA) <- make.names(paste(trimws(blat_penn.results.cDNA_header[1,]),trimws(blat_penn.results.cDNA_header[2,]),sep=""))
head(blat_penn.results.cDNA)
summary(blat_penn.results.cDNA)
```

### BLAT against genome
`r length(unique(blat_penn.results.genome$Qname))` of `r length(penn_contigs)` had at least one hit in S.lyc.

Number of contigs that have a unique hit:
```{r blat_penn_Stats1_genome, echo=FALSE}
sum(table(blat_penn.results.genome$Qname)==1)
```

### BLAT against cDNA

`r length(unique(blat_penn.results.cDNA$Qname))` of `r length(penn_contigs)` had at least one hit in S.lyc.

Number of contigs that have a unique hit:
```{r blat_penn_Stats1_cDNA, echo=FALSE}
sum(table(blat_penn.results.cDNA$Qname)==1)
```


## Filter contigs

For contigs that have multiple hits retain those where the best hit is substantially better than the next best hit.

### Genome
The number of contigs after filtering is:

```{r filter_penn_blat_genome, echo=FALSE}

blat_penn.results.genome.filtered <- lapply(unique(blat_penn.results.genome$Qname),function(contig) {#for each unique value of contig
  tmp.blat <- blat_penn.results.genome[blat_penn.results.genome$Qname==contig,] #make a temporary data frame with just the data from that contig
  if(nrow(tmp.blat)<2) return(tmp.blat) #if there is only a single row, return it
  tmp.score <- (tmp.blat$match - tmp.blat$mis.match) / tmp.blat$Qsize[1]
  tmp.blat <- tmp.blat[order(tmp.score,decreasing = TRUE),] #sort the temporary blat results
  tmp.score <- sort(tmp.score,decreasing = TRUE)
  if(tmp.score[1] > tmp.score[2]*1.1) { #check to see how much better the first hit is from the second
    return(tmp.blat[1,]) #return the best if it is much better than the second hit
  } else {
    return(tmp.blat[0,])#as.data.frame(matrix(NA,ncol=ncol(tmp.blat),nrow=1)))
  }
  }
  )

blat_penn.results.genome.filtered <- ldply(blat_penn.results.genome.filtered) #convert from list to data frame

nrow(blat_penn.results.genome.filtered)

# head(blat_penn.results.genome.filtered)
```

### cDNA
The number of contigs after filtering is:

```{r filter_penn_blat_cDNA, echo=FALSE}

blat_penn.results.cDNA.filtered <- lapply(unique(blat_penn.results.cDNA$Qname),function(contig) {#for each unique value of contig
  tmp.blat <- blat_penn.results.cDNA[blat_penn.results.cDNA$Qname==contig,] #make a temporary data frame with just the data from that contig
  if(nrow(tmp.blat)<2) return(tmp.blat) #if there is only a single row, return it
  tmp.score <- (tmp.blat$match - tmp.blat$mis.match) / tmp.blat$Qsize[1]
  tmp.blat <- tmp.blat[order(tmp.score,decreasing = TRUE),] #sort the temporary blat results
  tmp.score <- sort(tmp.score,decreasing = TRUE)
  if(tmp.score[1] > tmp.score[2]*1.1) { #check to see how much better the first hit is from the second
    return(tmp.blat[1,]) #return the best if it is much better than the second hit
  } else {
    return(tmp.blat[0,])#as.data.frame(matrix(NA,ncol=ncol(tmp.blat),nrow=1)))
  }
  }
  )

blat_penn.results.cDNA.filtered <- ldply(blat_penn.results.cDNA.filtered) #convert from list to data frame

nrow(blat_penn.results.cDNA.filtered)

# head(blat_penn.results.cDNA.filtered)
```

## Overlaps with lyc genes

### genome
Of these spen prime contigs, how many of them overlap a S.lyc gene in the genomic BLAT?

```{r contig_lyc_overlaps, echo=FALSE}
lyc.transcript.gr <- import("~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4_gene_models.gff3",feature.type="mRNA")

pen.contig.gr <- PSL2Granges(file="Assemblies_Sneo_Spen/pen_3prime_contigs_to_lyc_genome.psl")

pen.contig.gr.filter <- pen.contig.gr[blat_penn.results.genome.filtered$Qname]

pen.lyc.overlaps <- findOverlaps(pen.contig.gr.filter,lyc.transcript.gr) 

length(unique(queryHits(pen.lyc.overlaps)))
```

How many lyc genes are hit?
```{r contig_lyc_overlaps_genome_gene_gits, echo=FALSE}
length(unique(subjectHits(pen.lyc.overlaps)))
```

### cDNA
How many lyc genes are hit?

```{r contig_lyc_cDNA_gene_hits, echo=FALSE}
length(unique(blat_penn.results.cDNA.filtered$Tname))
```

## concatenate contigs based on blat

I want to compile all of the contigs that hit a gene into a single contig.  I should pay attention to strand and, ideally, order.  After looking at some of this I think that the best way is to do a cap3 assembly for contigs if there are multiple contigs.  Because I already know that these map to one gene I can use relaxed parameters in cap3.

```{r combine_contigs_blat_genome, eval=FALSE}
# for each subject hit
# grab all query hits
# then use cap3 to assemble if possible

# BLAT the cap3 results

# then string together based on cap3

#Strip extra characters from names
names(penn_assemblies) <- sub(" .*$","",names(penn_assemblies))   

setwd("Assemblies_Sneo_Spen")

#cleanup
if (file.exists("Spen_contigs_singletons_round2.fa")) file.remove("Spen_contigs_singletons_round2.fa")

system.time(for(s in unique(subjectHits(pen.lyc.overlaps))) {
  gene.name <- lyc.transcript.gr[s]$Name
  tmp.contig.index <- pen.lyc.overlaps[subjectHits(pen.lyc.overlaps)==s,]
  tmp.contig.names <- names(pen.contig.gr.filter[queryHits(tmp.contig.index)])
  tmp.contig.seq <- penn_assemblies[names(penn_assemblies) %in% tmp.contig.names]
  writeXStringSet(tmp.contig.seq,"tmp.fa") # add the gene name...
  system("cap3 tmp.fa -o 16 -s 251 -z 1 -p 66  -j 31",ignore.stdout = TRUE)
  
  system(paste("sed 's/>/>",gene.name,"|/' tmp.fa.cap.contigs tmp.fa.cap.singlets >> Spen_contigs_singletons_round2.fa",sep="")) 
})

file.remove(dir(pattern="tmp.fa"))

setwd("../")
```

##  Load new contigs
```{r Load_new}
Spen_contigs_round2.seq <- readDNAStringSet("Assemblies_Sneo_Spen/Spen_contigs_singletons_round2.fa")
Spen_contigs_round2.seq
```

## BLAT new contigs
```{r BLAT_new, engine='bash', eval=FALSE}
blat ~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4_cds.fasta Assemblies_Sneo_Spen/Spen_contigs_singletons_round2.fa -q=rna -t=dna -minMatch=4 -maxIntron=20000  Assemblies_Sneo_Spen/pen_3prime_contigs_V2_to_lyc_cDNA.psl
```

## Load new BLAT
```{r load_blat_new}
blat_penn_2.results.cDNA <- read.delim("Assemblies_Sneo_Spen/pen_3prime_contigs_V2_to_lyc_cDNA.psl",skip=5,stringsAsFactors = FALSE,header=FALSE)
blat_penn_2.results.cDNA_header <- read.delim("Assemblies_Sneo_Spen/pen_3prime_contigs_V2_to_lyc_cDNA.psl",skip=2,nrows=2,stringsAsFactors = FALSE,header=FALSE)
names(blat_penn_2.results.cDNA) <- make.names(paste(trimws(blat_penn_2.results.cDNA_header[1,]),trimws(blat_penn_2.results.cDNA_header[2,]),sep=""))
head(blat_penn_2.results.cDNA,15)
summary(blat_penn_2.results.cDNA)
```

## Filter new BLAT

filter to only keep results where the query hits the same gene again

```{r filter_new_BLAT}
blat_penn_2.results.cDNA.filtered <- blat_penn_2.results.cDNA[
  gsub("\\|[[:print:]]+$","",blat_penn_2.results.cDNA$Qname) == blat_penn_2.results.cDNA$Tname,]
nrow(blat_penn.results.cDNA)
nrow(blat_penn_2.results.cDNA.filtered)
length(unique(blat_penn_2.results.cDNA$Qname)) #contigs before filtering
length(unique(blat_penn_2.results.cDNA.filtered$Qname)) #contigs after filtering
length(unique(blat_penn_2.results.cDNA$Tname)) #genes before filtering
length(unique(blat_penn_2.results.cDNA.filtered$Tname)) #genes after filtering
```

```{r concatenate}
#for each gene
#get all contigs
#concatenate into a new DNA String
#add to new DNA string set

sfInit(parallel = TRUE,cpus = 2)
sfLibrary(Biostrings)
sfExport("blat_penn_2.results.cDNA.filtered","Spen_contigs_round2.seq")

# 2 minutes on 2 cpus
final_spen_contigs <- unlist(DNAStringSetList(sfLapply(unique(blat_penn_2.results.cDNA.filtered$Tname),function(g) {
  tmp.blat <- blat_penn_2.results.cDNA.filtered[blat_penn_2.results.cDNA.filtered$Tname==g,]
  tmp.blat <- tmp.blat[order(tmp.blat$Tstart),]
  concat.seq <- DNAStringSet(unlist(DNAStringSet(apply(tmp.blat,1,function(x) {
     tmp.seq <- Spen_contigs_round2.seq[[x[["Qname"]]]]
     if (x["strand"] == "-") reverseComplement(tmp.seq) else (tmp.seq)
   }))))
  names(concat.seq) <- paste(g,"Spen_Contig",sep="|")
  concat.seq
})))

sfStop()

writeXStringSet(final_spen_contigs,"Assemblies_Sneo_Spen/Spen_contigs_singletons_round3.fa")
   
```

```{r make_kallisto_index_round3, eval=FALSE, engine='bash'}
kallisto index -k 19 -i Assemblies_Sneo_Spen/Spen_3prime_contigs_round3.kai Assemblies_Sneo_Spen/Spen_contigs_singletons_round3.fa
```

## run kallisto
```{r run_kallisto_19_round3, eval=TRUE}
RunKallisto(fasta = "spen_threeprime_sim_1/sample_01_renamed.fastq.gz", index = "Assemblies_Sneo_Spen/Spen_3prime_contigs_round3.kai", dir = "spen_threeprime_sim_1/kallisto19_round3" )
```

## Load in mapped counts
```{r get_mapped_counts_kallisto19_round3 }
pen.mapped.counts <- GetMappedCounts(dir = "spen_threeprime_sim_1/kallisto19_round3", type="kallisto-transcripts")
```

# need to get working below here.

## _S. pennellii_
Compare known to mapped counts  
```{r compare_spen1_kallisto19, echo=FALSE, results='hide', fig.width=8, fig.height=5}
pen.count.comparison.1 <- CompareCounts(pen.known.counts, pen.mapped.counts, chrom.separate=FALSE, title = "S. pen. kallisto mapped to 3_prime_contigs 19nt kmer.  Kallisto counting")  
```

## Remove genes not in reference