---
title: "Spen_rnaseq_simulation"
author: "Julin Maloof"
date: "Jan 23, 2015"
output: html_document
---

I want to determine how well various aligners perform when mapping S. pennellii RNAseq reads to the Heinz genome.  The basic method will be to simulate reads from fasta files, map, and then compare results to reality.

I will start just with S.pen but will move on to differential expression analysis between M82 and S.pen.

I plan to compare

* stampy
* subread
* star
* bowtie

This script is for 3' DGE

### load libraries and helper scripts
```{r}
library(polyester)
library(Biostrings)
library(ggplot2)
library(Rsamtools)
library(rtracklayer)
library(Rsubread)
library(R.utils)
library(data.table)
setwd("~/Documents/Lab Notebook support/2015/rnaseq_simulations")
source("Rnaseq_sim_helper_functions.R")
```

## Run Mike's script to generate 3'UTR transcripts

Spen CDS fasta can be obtained from [SGN](ftp://ftp.solgenomics.net/genomes/Solanum_pennellii/annotations/)
slyc fasta is also from [SGN](ftp://ftp.sgn.cornell.edu/genomes/Solanum_lycopersicum/annotation/ITAG2.4_release/)

Only needs to be done once
```{r}
system(paste("~/git/extract-utr/extract-utr.pl",
             "--threeprime",
             "--cds_fa_file=/Users/jmaloof/Sequences/ref_genomes/S.pen/Spenn-v2-cds-annot.fa",
             "--gff_file=/Users/jmaloof/Sequences/ref_genomes/S.pen/spenn_v2.0_gene_models_annot.gff",
             "--genome_fa_file=/Users/jmaloof/Sequences/ref_genomes/S.pen/Spenn.fasta",
             "--output_fa_file=/Users/jmaloof/Sequences/ref_genomes/S.pen/Spenn-v2-3primeutr-250-250.fa",
             "--utr_length=250",
             "--gene_length=250"
))

system(paste("~/git/extract-utr/extract-utr.pl",
             "--threeprime",
             "--cds_fa_file=/Users/jmaloof/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4_cds.fasta",
             "--gff_file=/Users/jmaloof/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4_gene_models.gff3",
             "--genome_fa_file=/Users/jmaloof/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/S_lycopersicum_chromosomes.2.50.fa",
             "--output_fa_file=/Users/jmaloof/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4-threeprimeutr-250-250.fa",
             "--utr_length=250",
             "--gene_length=250"
))
```

## Modify ITAG gff
For read counting to work appropriately the Heinz GFF file needs to be modified to match the expected reads (i.e. +/- 250 from stop codon)


```{r}
library(rtracklayer)
heinz.gff <- readGFF("~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4_gene_models.gff3")
three.prime.df <- rbindlist(lapply(heinz.gff$Name[heinz.gff$type=="gene"], function(gene) {
  tmp <- as.data.frame(heinz.gff[grep(gene,heinz.gff$ID,fixed=T),])
  if (nlevels(factor(tmp$strand))>1) stop(paste("cannot detemine strand for gene",gene))
  if (all(tmp$strand=="+")) {
    stop.pos <- max(tmp$end[tmp$type=="CDS"])
  } else if (all(tmp$strand=="-")) {
    stop.pos <- min(tmp$start[tmp$type=="CDS"])
  } 
  result <- rbindlist(list(tmp[tmp$type=="gene",][1,],
                           tmp[tmp$type=="mRNA",][1,],
                           tmp[tmp$type=="exon",][1,]) )
  result$start <- stop.pos - 250 #this seems backwards for - strand, 
  #relative to the data.frame but it is required for converting to GRanges
  result$end <- stop.pos + 250
  as.data.frame(result)
}))
three.prime.GR <- makeGRangesFromDataFrame(three.prime.df,keep.extra.columns = TRUE)
export(three.prime.GR,"~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4_three_prime_250_250.gff3","gff3")
```


## Simluate DGE reads (simple version)

Load Spen and Slyc 3' UTR  

```{r load_Spen, eval=FALSE}
pen.transcripts <-  readDNAStringSet("~/Sequences/ref_genomes/S.pen/Spenn-v2-3primeutr-250-250.fa")
head(pen.transcripts)
names(pen.transcripts[1:10])
lyc.transcripts <- readDNAStringSet("~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4-threeprimeutr-250-250.fa")
head(lyc.transcripts)
names(lyc.transcripts)[1:10]
```

Reformat names to get rid of description and add Slyc reference
```{r reformat_gene_names, eval=FALSE}
orthologs <- read.delim("Ortho_v2.txt",header=FALSE) #From Tony Bolger
head(orthologs)
names(orthologs) <- c("PEN","LYC")
orthologs$PEN <- sub("t","",orthologs$PEN)
orthologs$PEN.LYC <- paste(orthologs$PEN,orthologs$LYC,sep="_")
names(pen.transcripts) <- substr(names(pen.transcripts),1,16)
names(pen.transcripts) <- ifelse(!is.na(match(names(pen.transcripts),orthologs$PEN)),
                                 orthologs$PEN.LYC[match(names(pen.transcripts),orthologs$PEN)],
                                 names(pen.transcripts))
pen.transcripts

names(lyc.transcripts) <- substr(names(lyc.transcripts),1,18)
names(lyc.transcripts) <- ifelse(!is.na(match(names(lyc.transcripts),orthologs$LYC)),
                                 orthologs$PEN.LYC[match(names(lyc.transcripts),orthologs$LYC)],
                                 names(lyc.transcripts))
pen.gene.length <- as.integer(nchar(pen.transcripts))
names(pen.gene.length) <- names(pen.transcripts)
lyc.gene.length <- as.integer(nchar(lyc.transcripts))
names(lyc.gene.length) <- names(lyc.transcripts)
```

Lets limit ourselves to genes where orthology is clear:
```{r reduce_data, eval=FALSE}
pen.transcripts.small <- pen.transcripts[nchar(names(pen.transcripts))>16]
writeXStringSet(pen.transcripts.small,file="~/Sequences/ref_genomes/S.pen/Spenn-v2-3primeutr-250-250-orthologs-only.fa")
lyc.transcripts.small <- lyc.transcripts[nchar(names(lyc.transcripts)) > 18]
writeXStringSet(lyc.transcripts.small,file="~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4-threeprimeutr-250-250-orthlogs-only.fa")
```

Simulate PEN RNAseq reads
```{r simulate_pen_1, eval=FALSE}
#15 minutes for 5M reads
system.time(
  simulate_experiment(fasta="~/Sequences/ref_genomes/S.pen/Spenn-v2-3primeutr-250-250-orthologs-only.fa",
                      outdir="spen_threeprime_sim_1",
                      num_reps=1,
                      readlen=45,
                      #line below gives ~ 5M reads, scaled by transcript length
                      reads_per_transcript = round(width(pen.transcripts.small) * 
                                                     5e6/ sum(width(pen.transcripts.small) )),
                      paired=FALSE,
                      error_model="illumina5",
                      bias="rnaf"
  )
)
system("sed 's!/!;!' spen_threeprime_sim_1/sample_01.fasta > spen_threeprime_sim_1/sample_01_renamed.fasta") #this allows original gene_names to be kept
system("rm spen_threeprime_sim_1/sample_01.fasta")
system("gzip spen_threeprime_sim_1/sample_01_renamed.fasta")
```

Simulate LYC RNAseq reads
```{r simulate_lyc_1, eval=FALSE}
system.time(
  simulate_experiment(fasta="~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4-threeprimeutr-250-250-orthlogs-only.fa",
                      outdir="slyc_threeprime_sim_1",
                      num_reps=1,
                      readlen=45,
                      #line below gives ~ 5M reads, scaled by transcript length
                      reads_per_transcript = round(width(lyc.transcripts.small) * 
                                                     5e6/ sum(width(pen.transcripts.small) )),
                      paired=FALSE,
                      error_model="illumina5",
                      bias="rnaf"
  )
)
system("sed 's!/!;!' slyc_threeprime_sim_1/sample_01.fasta > slyc_threeprime_sim_1/sample_01_renamed.fasta") #this allows original gene_names to be kept
system("rm slyc_threeprime_sim_1/sample_01.fasta")
system("gzip slyc_threeprime_sim_1/sample_01_renamed.fasta")
```

### calculate known counts

use external python script
```{r known_counts}
pen.known.counts <- read.table(pipe("./CountGene.py spen_threeprime_sim_1/sample_01_renamed.fasta.gz"),col.names=c("id","count"),stringsAsFactors = FALSE)
pen.known.counts$lyc.id <- substr(pen.known.counts$id,18,33)
head(pen.known.counts)

lyc.known.counts <- read.table(pipe("./CountGene.py slyc_threeprime_sim_1/sample_01_renamed.fasta.gz"),col.names=c("id","count"),stringsAsFactors = FALSE)
lyc.known.counts$lyc.id <- substr(lyc.known.counts$id,18,33)
head(lyc.known.counts)
```

## Map with star

### Create genome file
Only needs to be done once  
This is a shell command  
Takes about 25 minutes  
```{r create_genome_file, eval=FALSE, engine='bash'}
gffread -T ITAG2.4_gene_models.gff3 -o ITAG2.4_gene_models.gtf
STAR --runThreadN 6 --runMode genomeGenerate --genomeDir ~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/STAR_genome --genomeFastaFiles ~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/S_lycopersicum_chromosomes.2.50.fa --sjdbGTFfile ~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4_gene_models.gtf
```

### Create parameters files
```{r STAR_params_1, eval=FALSE, engine='bash'}
cat > STAR.params.whitney.1
genomeDir /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/jmaloof/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/STAR_genome
outSAMtype BAM SortedByCoordinate
outSAMunmapped Within

cat > STAR.params.royce.1
genomeDir /Users/jmaloof/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/STAR_genome
outSAMtype BAM SortedByCoordinate
outSAMunmapped Within
```

### Run it!
```{r RunStar_1, eval=FALSE}
RunStar(fasta="sample_01_renamed.fasta.gz",prefix = "STAR_1",dir="spen_threeprime_sim_1",param.file = "../STAR.params.whitney.1")
RunStar(fasta="sample_01_renamed.fasta.gz",prefix = "STAR_1",dir="slyc_threeprime_sim_1",param.file = "../STAR.params.whitney.1")
```
__note__ I needed to use STAR2.5 for this to work.

## Evaluate results

We want to do:

* correlation between known counts and found counts
* how many reads map to the wrong gene?
* how many reads do not map?
* something about splice junctions

### Load in mapped counts
```{r get_mapped_counts }
#using the STAR built-in counter
pen.mapped.counts.STAR <- GetMappedCounts(prefix = "STAR_1", dir = "spen_threeprime_sim_1")
lyc.mapped.counts.STAR <- GetMappedCounts(prefix = "STAR_1", dir = "slyc_threeprime_sim_1")
#using featureCounts from Rsubread
pen.mapped.counts <- GetMappedCounts(prefix = "STAR_1",
                                     dir = "spen_threeprime_sim_1",
                                     type="STAR-featureCounts",
                                    gff="~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4_three_prime_250_250.gff3")

lyc.mapped.counts <- GetMappedCounts(prefix = "STAR_1",
                                     dir = "slyc_threeprime_sim_1",
                                     type="STAR-featureCounts",
                                     gff="~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4_three_prime_250_250.gff3")
```

### Comparison

```{r compare_spen1}
pen.count.comparison.1.STAR <- CompareCounts(pen.known.counts, pen.mapped.counts.STAR, title = "S. pen.  STAR mapped.  Default parameters")
lyc.count.comparison.1.STAR <- CompareCounts(lyc.known.counts, lyc.mapped.counts.STAR, title = "S. lyc.  STAR mapped.  Default parameters")
pen.count.comparison.1 <- CompareCounts(pen.known.counts, pen.mapped.counts, title = "S. pen.  STAR mapped.  Default parameters.  FeatureCounts counting")
lyc.count.comparison.1 <- CompareCounts(lyc.known.counts, lyc.mapped.counts, title = "S. lyc.  STAR mapped.  Default parameters FeatureCounts counting")

```


## Map with STAR to Heinz 3' DGE

### Create "genome" file
Only needs to be done once  
This is a shell command  
Takes about 25 minutes  
```{r create_genome_file, eval=FALSE, engine='bash'}
STAR --runThreadN 2 --runMode genomeGenerate --genomeDir ~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/STAR_3prime --genomeFastaFiles ~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4-threeprimeutr-250-250.fa
```

### Create parameters files
```{r STAR_params_1, eval=FALSE, engine='bash'}
cat > STAR.params.whitney.3prime
genomeDir /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/jmaloof/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/STAR_3prime
outSAMtype BAM SortedByCoordinate
outSAMunmapped Within

cat > STAR.params.royce.3prime
genomeDir /Users/jmaloof/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/STAR_3prime
outSAMtype BAM SortedByCoordinate
outSAMunmapped Within
```

### Run it!
```{r RunStar_1, eval=FALSE}
RunStar(fasta="sample_01_renamed.fasta.gz",prefix = "STAR_1",dir="spen_threeprime_sim_1",param.file = "../STAR.params.whitney.1")
RunStar(fasta="sample_01_renamed.fasta.gz",prefix = "STAR_1",dir="slyc_threeprime_sim_1",param.file = "../STAR.params.whitney.1")
```


### Below not run on 3'DGE stuff yet


Investigate...
Lets take a look at the bam file and extract reads from a few genes with 0 mapped reads.  Skip chromosome 0
```{r}
reads <- scanBam(file="spen_threeprime_sim_1/STAR_1Aligned.sortedByCoord.out.bam",
                 param = ScanBamParam(what=scanBamWhat(),
                                      tag=c("NH","HI","AS","nM")))[[1]]
#note: DO NOT directly convert reads to a data frame.  Uses lots of memory.  instead remove the very large reads column.  See below.

reads <- c(reads[names(reads)!="tag"],reads$tag) #unpack each element of reads$tag and add it to reads

#add an element indicating whether or not the read came from a gene with no reads mapping to the pen version
genes.zero.map <- pen.count.comparison.1$id[pen.count.comparison.1$count.known>0 & pen.count.comparison.1$count.mapped==0]
reads$gene <- unlist(strsplit(reads$qname,";"))[c(FALSE,TRUE,FALSE,FALSE)]
lapply(reads, head)
reads$pen.zero.map <- reads$gene %in% genes.zero.map

table(reads$flag,reads$pen.zero.map)
table(reads$NH,reads$pen.zero.map) #number of hits

reads.df <- with(reads,data.frame(qname,rname,pos,qwidth,flag,cigar,NH,HI,AS,nM,gene,pen.zero.map,stringsAsFactors=FALSE))
reads.df$gene.lyc <- substr(reads.df$gene,18,38)

reads.df$multimap <- reads.df$NH>1

reads.df$pen.gene.length <- pen.gene.length[reads.df$gene]
reads.df$lyc.gene.length <- lyc.gene.length[reads.df$gene]

reads.df$pen.lyc.gene.length.dif <- reads.df$pen.gene.length - reads.df$lyc.gene.length

reads.df$pen.lyc.gene.length.ratio <- reads.df$pen.gene.length / reads.df$lyc.gene.length

ggplot(reads.df,aes(x=as.factor(NH),y=nM,group=NH)) + geom_point(stat="sum") + scale_size(breaks=seq(0,1,.1),range=c(1,15))

ggplot(reads.df,aes(x=as.factor(NH),y=pen.gene.length,fill=pen.zero.map)) + geom_violin()

ggplot(reads.df,aes(x=as.factor(NH),y=lyc.gene.length,fill=pen.zero.map)) + geom_violin()

ggplot(reads.df,aes(x=as.factor(NH),y=pen.gene.length)) + geom_violin()

ggplot(reads.df,aes(x=as.factor(NH),y=pen.lyc.gene.length.ratio,fill=multimap)) + geom_violin()

ggplot(reads.df,aes(x=pen.zero.map,y=pen.gene.length,fill=multimap)) + geom_boxplot()

ggplot(reads.df,aes(x=pen.zero.map,y=pen.lyc.gene.length.ratio,fill=multimap)) + geom_violin()

#maybe better to look at this gene_length thing on a per gene basis

ggplot(pen.count.comparison.1,aes(x=gene.length.diff.pen.lyc,y=count.ratio.known.mapped)) + geom_point(alpha=0.1) + scale_y_log10() + scale_x_log10(breaks=c(1,10,100,200,300,500,1000))

ggplot(pen.count.comparison.1,aes(x=gene.length.ratio.pen.lyc,y=count.ratio.known.mapped))  + scale_y_log10() + scale_x_log10(breaks=c(.1,.5,seq(1,2,.2),3,5,10)) + geom_hline(yintercept=1,color="blue") + geom_vline(xintercept=1,color="blue") + geom_point(alpha=.1)

#reasonable cutoffs might be a gene.length.ratio > 1.3 or a gene length difference of > 300

pen.count.comparison.1.cutoff1.3 <- pen.count.comparison.1[pen.count.comparison.1$gene.length.ratio.pen.lyc < 1.3,]
nrow(pen.count.comparison.1)
nrow(pen.count.comparison.1.cutoff1.3)
with(pen.count.comparison.1.cutoff1.3,cor(count.known,count.mapped))

pen.count.comparison.1.cutoff300 <- pen.count.comparison.1[pen.count.comparison.1$gene.length.diff.pen.lyc < 300,]
nrow(pen.count.comparison.1)
nrow(pen.count.comparison.1.cutoff300)
with(pen.count.comparison.1.cutoff300,cor(count.known,count.mapped))

pen.count.comparison.1.cutoff300.1.3 <- pen.count.comparison.1[pen.count.comparison.1$gene.length.diff.pen.lyc < 300 & pen.count.comparison.1$gene.length.ratio.pen.lyc < 1.3,]
nrow(pen.count.comparison.1)
nrow(pen.count.comparison.1.cutoff300.1.3)
with(pen.count.comparison.1.cutoff300.1.3,cor(count.known,count.mapped))

length.cutoff.plot.data <- as.data.frame(t(sapply(seq(0,max(pen.count.comparison.1$gene.length.diff.pen.lyc),50),function(x) 
  with(
    pen.count.comparison.1[pen.count.comparison.1$gene.length.diff.pen.lyc <= x ,],
    c(cutoff=x,genes=length(count.known),correlation=cor(count.known,count.mapped))))))
length.cutoff.plot.data$type <- "difference"


ratio.cutoff.plot.data <- as.data.frame(t(sapply(seq(1,max(pen.count.comparison.1$gene.length.ratio.pen.lyc),.1),function(x) 
  with(
    pen.count.comparison.1[pen.count.comparison.1$gene.length.ratio.pen.lyc <= x ,],
    c(cutoff=x,genes=length(count.known),correlation=cor(count.known,count.mapped))))))
ratio.cutoff.plot.data$type <- "ratio"

combined.cutoff.plot.data <-rbind(length.cutoff.plot.data,ratio.cutoff.plot.data)

ggplot(combined.cutoff.plot.data,aes(x=genes,y=correlation,color=type)) + geom_line() 

ggplot(length.cutoff.plot.data,aes(x=genes,y=correlation,label=cutoff)) + geom_line() + geom_text(aes(label=cutoff),size=2.5,vjust=-1,hjust=-.3) 
```

how many reads map to the wrong gene?

note: there is ranges data for each read mapping.  So I can compare that to the GTF ranges.
```{r mismapping}
lyc.transcript.gr <- import("~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4_gene_models.gff3",feature.type="mRNA")
reads.df.map <- na.omit(reads.df)
reads.gr <- with(reads.df.map,GRanges(seqnames=rname,
                                      ranges = IRanges(start=pos,
                                                       width = qwidth, #note that this is not precisely correct because of introns,etc
                                                       names=qname)))
hits <- findOverlaps(reads.gr,lyc.transcript.gr,ignore.strand=TRUE)
sum(reads.df.map$gene.lyc[queryHits(hits)] != lyc.transcript.gr$Name[subjectHits(hits)]) / length(hits)
reads.df.map.single <- reads.df.map[reads.df.map$NH==1,]
reads.gr.single <- with(reads.df.map.single,GRanges(seqnames=rname,
                                                    ranges = IRanges(start=pos,
                                                                     width = qwidth, #note that this is not precisely correct because of introns,etc
                                                                     names=qname)))
hits.single <- findOverlaps(reads.gr.single,lyc.transcript.gr,ignore.strand=TRUE)
sum(reads.df.map.single$gene.lyc[queryHits(hits.single)] != lyc.transcript.gr$Name[subjectHits(hits.single)]) / length(hits.single)

reads.df.map.single.overlap <- reads.df.map.single[queryHits(hits.single),] #only keep reads that overlap a transcript

reads.df.map.single.overlap$gene.mapped.to <- lyc.transcript.gr$Name[subjectHits(hits.single)]

#remove chromosome zero

reads.df.map.single.overlap <- reads.df.map.single.overlap[!grepl("Solyc00g",paste(reads.df.map.single.overlap$gene.lyc,reads.df.map.single.overlap$gene.mapped.to)),] #not very different

with(reads.df.map.single.overlap,sum(gene.lyc!=gene.mapped.to)/length(gene.lyc))

#are reads represented more than once?
length(unique(reads.df.map.single.overlap$qname))/nrow(reads.df.map.single.overlap)
#yes.  So reads are overlapping with multiple genes.  Presumably this is due to overlaping genes in the annotation.

reads.df.map.single.overlap$chrom.lyc <- substr(reads.df.map.single.overlap$gene.lyc,6,8)
reads.df.map.single.overlap$chrom.mapped.to <- substr(reads.df.map.single.overlap$gene.mapped.to,6,8)


reads.df.map.single.overlap$mismap <- reads.df.map.single.overlap$gene.lyc != reads.df.map.single.overlap$gene.mapped.to
head(reads.df.map.single.overlap,50)
head(reads.df.map.single.overlap[reads.df.map.single.overlap$mismap,c("gene.lyc","gene.mapped.to")],50)
#yes the overlapping annotaiton seems to explain some of it.  Can check more
reads.df.map.single.overlap <- within(reads.df.map.single.overlap, mapnear <- 
                                        (abs(as.numeric(substr(gene.lyc,9,14)) - as.numeric(substr(gene.mapped.to,9,14))) < 40) & 
                                        chrom.lyc == chrom.mapped.to)
head(reads.df.map.single.overlap[reads.df.map.single.overlap$mismap,c("gene.lyc","gene.mapped.to","mapnear")],100)
with(reads.df.map.single.overlap,sum(mapnear & mismap) / sum(mismap))

#so if we get rid of the "mapnear" mismaps how many mismaps are left?
sum(reads.df.map.single.overlap$mismap[!reads.df.map.single.overlap$mapnear]) / nrow(reads.df.map.single.overlap)

#save.image("rnaseqSim.Rdata")
save(list=grep("reads",ls(),invert = TRUE,value=T),file="rnaseqSim_no_read_data.Rdata")
```

Next steps: some comparative genomic alignments for the genes that aren't behaving well.  (ideas: BLAT produces PSL files and IGV can read those.  Or use STAR.  Or use BLAST and the blast2sam).

```{r BLAT install, eval=FALSE, engine='bash'}
brew install blat
```

```{r run BLAT, engine='bash', eval=FALSE}
blat ~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/S_lycopersicum_chromosomes.2.50.fa -q=rna -t=dna -minMatch=4 ~/Sequences/ref_genomes/S.pen/Spenn-v2-cds-annot_orthlogs_only.fa -maxIntron=20000 penCDS_to_lycGENME.psl
#setting minMatch to 4 reduces number of overlaps from ~6300 to 3300
```

Map _real_ reads as comparision
```{r map_real,eval=FALSE}
RunStar(fasta = "/Volumes/temp-o-matic/temp_SolRAID_restore/Solexa_runs_Data/Filtered_fastq_by_seedling_exp/merged_files/PEN.seedlings.fq",
        prefix = "STAR_1",
        dir = "spen_real",
        n=12,
        param.file = "../STAR.params.whitney.1")
```


Now look in IGV (Note: Need to use IGVtools to sort and index the .psl file)

```{r}
head(reads.df.map.single.overlap[reads.df.map.single.overlap$mismap,c("qname","gene.lyc","gene.mapped.to","mapnear")],100)
sort(table(reads.df.map.single.overlap$gene.lyc[reads.df.map.single.overlap$mismap]))
```

### 5 examples of genes with mismapping

Solyc02g090930.2.1 (12): Tandem Repeat with Solyc02g090940.2.1
Solyc04g072480.2.1 (30): Nothing obviously strange
Solyc03g051930.2.1 (70): PEN gene spans two LYC genes
Solyc02g079850.2.1 (150): PEN gene spans two LYC genes
Solyc01g110960.2.1 (497): Much Confusion

### 5 examples of genes with zero mapping

```{r}
pen.count.comparison.1.cutoff300.1.3[pen.count.comparison.1.cutoff300.1.3$count.ratio==Inf,c("lyc.id","count.known","gene.length.ratio.pen.lyc")]
```

Solyc01g010050.2: Two overlapping LYC transcripts (note this is fixed when using FeatureCounts from subread)
Solyc01g060020.2: Multiple overlapping LYC transcripts
Solyc05g023780.1: Repetitive region
Solyc08g015710.1: Very small region of homology
Solyc11g056540.1: BLAT did not find a match

Many of these issues could be dealt with by
1. Eliminating genes with overlapping LYC to LYC or PEN to LYC transcripts. 
2. Using the BLAT results to eliminate transcripts with poor homology.

```{r}
lyc.lyc.overlaps <- findOverlaps(lyc.transcript.gr,lyc.transcript.gr) #all overlaps including self on self
lyc.lyc.overlaps <- lyc.lyc.overlaps[queryHits(lyc.lyc.overlaps)!=subjectHits(lyc.lyc.overlaps),] #limit to non-self overlaps
lyc.lyc.overlaps <- substr(unique(mcols(lyc.transcript.gr)[queryHits(lyc.lyc.overlaps),"Name"]),1,16)

pen.transcript.gr <- PSL2Granges(file="penCDS_to_lycGENME.psl")
pen.pen.overlaps <- findOverlaps(pen.transcript.gr,pen.transcript.gr) #all overlaps including self on self

#need additional step to limit to non-self overlaps because BLAT returns multiple hits in the same region.  Make sure that the overlap is with the same gene
pen.pen.overlaps <- pen.pen.overlaps[names(pen.transcript.gr)[queryHits(pen.pen.overlaps)]!=names(pen.transcript.gr)[subjectHits(pen.pen.overlaps)]] 

pen.pen.overlaps <- unique(mcols(pen.transcript.gr)[queryHits(pen.pen.overlaps),"lyc.ID"]) 
```

Now if we eliminate these genes how are our count comparisons?
```{r}
pen.count.comparison.1$flag <- "none"
pen.count.comparison.1 <- within(pen.count.comparison.1,{
  flag[pen.count.comparison.1$lyc.id %in% lyc.lyc.overlaps] <- "lyc.lyc.overlap"
  flag[pen.count.comparison.1$lyc.id %in% pen.pen.overlaps] <- "pen.pen.overlap"
}
)
with(pen.count.comparison.1,cor(count.known,count.mapped))
with(pen.count.comparison.1[pen.count.comparison.1$flag=="none",],cor(count.known,count.mapped))
```


