---
title: "Spen_rnaseq_simulation"
author: "Julin Maloof"
date: "November 29, 2015"
output: html_document
---

I want to determine how well various aligners perform when mapping S. pennellii RNAseq reads to the Heinz genome.  The basic method will be to simulate reads from fasta files, map, and then compare results to reality.

I will start just with S.pen but will move on to differential expression analysis between M82 and S.pen.

I plan to compare

* stampy
* subread
* star
* bowtie

## Simluate reads (simple version)

Use the [Polyester package.](http://bioconductor.org/packages/release/bioc/vignettes/polyester/inst/doc/polyester.html)  Also see the [Polyester paper.](http://bioinformatics.oxfordjournals.org/content/31/17/2778)

Installation.  Only needs to be run once per computer.
```{r install_polyester, eval=FALSE}
source("http://bioconductor.org/biocLite.R")
biocLite("polyester")
```

Load Spen transcripts  
Spen CDS fasta can be obtained from [SGN](ftp://ftp.solgenomics.net/genomes/Solanum_pennellii/annotations/)
```{r load_Spen}
library(polyester)
library(Biostrings)
setwd("~/Documents/Lab Notebook support/2015/rnaseq_simulations")
pen.transcripts <-  readDNAStringSet("~/Sequences/ref_genomes/S.pen/Spenn-v2-cds-annot.fa")
head(pen.transcripts)
names(pen.transcripts[1:10])
```

Reformat names to get rid of description and add Slyc reference
```{r }
orthologs <- read.delim("Ortho_v2.txt",header=FALSE) #From Tony Bolger
head(orthologs)
names(orthologs) <- c("PEN","LYC")
orthologs$PEN <- sub("t","",orthologs$PEN)
head(orthologs)
orthologs$PEN.LYC <- paste(orthologs$PEN,orthologs$LYC,sep="_")
names(pen.transcripts) <- substr(names(pen.transcripts),1,16)
names(pen.transcripts) <- ifelse(!is.na(match(names(pen.transcripts),orthologs$PEN)),
                                 orthologs$PEN.LYC[match(names(pen.transcripts),orthologs$PEN)],
                                 names(pen.transcripts))
pen.transcripts
```

Lets limit ourselve to genes where orthology is clear:
```{r reduce_data}
pen.transcripts.small <- pen.transcripts[nchar(names(pen.transcripts))>16]
writeXStringSet(pen.transcripts.small,file="~/Sequences/ref_genomes/S.pen/Spenn-v2-cds-annot_orthlogs_only.fa")
```


Simulate PEN RNAseq reads
```{r simulate_pen_1}
#3 minutes for 1M reads
#15 minutes for 5M reads
system.time(
  simulate_experiment(fasta="~/Sequences/ref_genomes/S.pen/Spenn-v2-cds-annot_orthlogs_only.fa",
                      outdir="spen_sim_1",
                      num_reps=1,
                      readlen=45,
                      #line below gives ~ 5M reads, scaled by transcript length
                      reads_per_transcript = round(width(pen.transcripts.small) * 
                                                     5e6/ sum(width(pen.transcripts.small) )),
                      paired=FALSE,
                      error_model="illumina5",
                      bias="rnaf",
                      transcriptid=names(pen.transcripts.small)
  )
)
system("sed 's!/!;!' spen_sim_1/sample_01.fasta > spen_sim_1/sample_01_renamed.fasta") #this allows original gene_names to be kept
```

## Map with star

### Installation (only needs to be done once)

brew install rna-star
brew install gcc5 

### Create genome file
Only needs to be done once  
This is a shell command  
Takes about 25 minutes  
```{r eval=FALSE, engine='bash'}
gffread -T ITAG2.4_gene_models.gff3 -o ITAG2.4_gene_models.gtf
STAR --runThreadN 6 --runMode genomeGenerate --genomeDir ~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/STAR_genome --genomeFastaFiles ~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/S_lycopersicum_chromosomes.2.50.fa --sjdbGTFfile ~/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/ITAG2.4_gene_models.gtf
```

### Create parameters files
```{r eval=FALSE, engine='bash'}
cat > STAR.params.1
runThreadN 12
genomeDir /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/jmaloof/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/STAR_genome
twopassMode Basic
outSAMtype BAM SortedByCoordinate
outSAMunmapped Within
outFilterMismatchNoverLmax 0.1
outFilterIntronMotifs RemoveNoncanonical
quantMode GeneCounts

```

### Run it!

__note__ I needed to use STAR2.5 for this to work.

```{r eval=FALSE, engine='bash'}
cd spen_sim_1
STAR --parametersFiles ../STAR.params.1 --readFilesIn ./sample_01_renamed.fasta --outFileNamePrefix STAR_1 
samtools index STAR_1Aligned.sortedByCoord.out.bam
```

## Evaluate results

We want to do:

* correlation between known counts and found counts
* how many reads map to the wrong gene?
* how many reads do not map?
* something about splice junctions

### calculate known counts

A few options are presented below

use external python script
```{r}
known_counts <- read.table(pipe("./CountGene.py spen_sim_1/sample_01.fasta"),col.names=c("id","count"),stringsAsFactors = FALSE)
summary(known_counts)
head(known_counts)
```


If you are not concerned about memory usage:
(5M reads uses 4GB)
```{r}
library(Biostrings)
system.time({
  reads <- readDNAStringSet("spen_sim_1/sample_01.fasta")
    known_counts <- as.data.frame(
      table(
        regmatches(names(reads),regexpr("Sopen.+Solyc[0-9]{2}g[0-9\\.]+",names(reads))),
        dnn="id"),
      responseName="count",
      stringsAsFactors = FALSE)
})
```

Memory-friendly way to count gene occurences...
2-3 minutes for 5M reads
```{r}
system.time({
  known_counts <- list()
  reads <- file("spen_sim_1/sample_01.fasta")
  open(reads)
  repeat {
    lines <- readLines(reads,n=100000)
    genes <- as.data.frame(
      table(
        regmatches(lines,regexpr("Sopen.+Solyc[0-9]{2}g[0-9\\.]+",lines)),
        dnn="id"),
      responseName="count",
      stringsAsFactors = FALSE)
    matches <- genes$id %in% names(known_counts)
    known_counts[genes$id[!matches]] <- 0 #initiate an entry for new genes
    known_counts[genes$id] <- unlist(known_counts[genes$id]) + genes$count
    if(length(lines)==0) break
  }
  close(reads)    
})
```

### Load in mapped counts
```{r }
mapped_counts <- read.delim("spen_sim_1/STAR_1ReadsPerGene.out.tab",header = FALSE, stringsAsFactors = FALSE)
head(mapped_counts)
mapped_counts <- mapped_counts[,1:2] # 3 and 4 are directional counts which we don't care about
colnames(mapped_counts) <- c("ID","count")
mapped_counts$ID <- sub("gene:","",mapped_counts$ID,fixed=TRUE)
```

### Comparison
```{r}
#check that I can use substring
unique(nchar(known_counts$id)) #all are the same length
known_counts$lyc_id <- substr(known_counts$id,18,33)
head(known_counts)
count_comparison <- merge(known_counts,mapped_counts,by.x="lyc_id",by.y="ID",suffixes=c(".known",".mapped"))
head(count_comparison)
cor(count_comparison$count.known,count_comparison$count.mapped)
plot(count_comparison$count.known,count_comparison$count.mapped)
```
Not so good!

how much is chromosome zero?
```{r}
count_comparison$chrom <- substr(count_comparison$lyc_id,6,7)
head(count_comparison)
library(ggplot2)
qplot(x=count.known,y=count.mapped,data=count_comparison) + facet_wrap(~ chrom)
```
Not limited to chrom zero

Investigate...
Lets take a look at the bam file and extract reads from a few genes with 0 mapped reads.  Skip chromosome 0
```{r}
head(count_comparison[count_comparison$chrom=="01" & count_comparison$count.mapped==0,],n=20)
library(Rsamtools)
reads <- scanBam(file="spen_sim_1/STAR_1Aligned.sortedByCoord.out.bam",
                 param = ScanBamParam(what=scanBamWhat(),
                                      tag=c("NH","HI","AS","nM")))[[1]]
#note: DO NOT convert reads to a data frame.  Uses lots of memory.
reads <- c(reads[names(reads)!="tag"],reads$tag)
lapply(reads, head)
#pull the reads that should have mapped out into new list
zero_map_genes <- count_comparison$id[count_comparison$count.known>0 & count_comparison$count.mapped==0]
reads$gene <- unlist(strsplit(reads$qname,";"))[c(FALSE,TRUE,FALSE,FALSE)]
lapply(reads, head)
reads_zero <- lapply(reads,function(x) x[reads$gene %in% zero_map_genes])
lapply(reads_zero,head,20)
table(reads_zero$flag)
hist(reads_zero$NH) #number of hits
hist(reads$NH)
#so a fair number of the reads from genes with zero counts are multi-mappers.

#what about the other ones

reads_zero_1hit <- lapply(reads_zero,function(x) x[reads_zero$NH==1])
lapply(reads_zero_1hit,head,10)
#remember: allow non-cannonical...
```

### Create parameters files
```{r eval=FALSE, engine='bash'}
cat > STAR.params.2
runThreadN 12
genomeDir /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/jmaloof/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/STAR_genome
twopassMode Basic
outSAMtype BAM SortedByCoordinate
outSAMunmapped Within
outFilterMismatchNoverLmax 0.1
quantMode GeneCounts
```

### Run it!

__note__ I needed to use STAR2.5 for this to work.

```{r eval=FALSE, engine='bash'}
cd spen_sim_1
STAR --parametersFiles ../STAR.params.2 --readFilesIn ./sample_01_renamed.fasta --outFileNamePrefix STAR_2 
samtools index STAR_2Aligned.sortedByCoord.out.bam
```

### Load in mapped counts
```{r }
mapped_counts.2 <- read.delim("spen_sim_1/STAR_2ReadsPerGene.out.tab",header = FALSE, stringsAsFactors = FALSE)
head(mapped_counts.2)
mapped_counts.2 <- mapped_counts.2[,1:2] # 3 and 4 are directional counts which we don't care about
colnames(mapped_counts.2) <- c("ID","count")
mapped_counts.2$ID <- sub("gene:","",mapped_counts.2$ID,fixed=TRUE)
```

### Comparison
```{r}
#check that I can use substring
unique(nchar(known_counts$id)) #all are the same length
known_counts$lyc_id <- substr(known_counts$id,18,33)
head(known_counts)
count_comparison.2 <- merge(known_counts,mapped_counts.2,by.x="lyc_id",by.y="ID",suffixes=c(".known",".mapped"))
head(count_comparison.2)
cor(count_comparison.2$count.known,count_comparison.2$count.mapped)
plot(count_comparison.2$count.known,count_comparison.2$count.mapped)
```
Not so good!

how much is chromosome zero?
```{r}
count_comparison.2$chrom <- substr(count_comparison.2$lyc_id,6,7)
head(count_comparison.2)
library(ggplot2)
qplot(x=count.known,y=count.mapped,data=count_comparison.2) + facet_wrap(~ chrom)
```
Not limited to chrom zero

Investigate...
Lets take a look at the bam file and extract reads from a few genes with 0 mapped reads.  Skip chromosome 0
```{r}
head(count_comparison.2[count_comparison.2$chrom=="01" & count_comparison.2$count.mapped==0,],n=20)
library(Rsamtools)
reads <- scanBam(file="spen_sim_1/STAR_2Aligned.sortedByCoord.out.bam",
                 param = ScanBamParam(what=scanBamWhat(),
                                      tag=c("NH","HI","AS","nM")))[[1]]
#note: DO NOT convert reads to a data frame.  Uses lots of memory.
reads <- c(reads[names(reads)!="tag"],reads$tag)
lapply(reads, head)
#pull the reads that should have mapped out into new list
zero_map_genes <- count_comparison.2$id[count_comparison.2$count.known>0 & count_comparison.2$count.mapped==0]
reads$gene <- unlist(strsplit(reads$qname,";"))[c(FALSE,TRUE,FALSE,FALSE)]
lapply(reads, head)
reads_zero <- lapply(reads,function(x) x[reads$gene %in% zero_map_genes])
lapply(reads_zero,head,20)
table(reads_zero$flag)
hist(reads_zero$NH) #number of hits
hist(reads$NH)
#so a fair number of the reads from genes with zero counts are multi-mappers.

#what about the other ones

reads_zero_1hit <- lapply(reads_zero,function(x) x[reads_zero$NH==1])
lapply(reads_zero_1hit,head,10)
```

### Create parameters files
```{r eval=FALSE, engine='bash'}
cat > STAR.params.3
runThreadN 12
genomeDir /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/jmaloof/Sequences/ref_genomes/tomato/ITAG2.4_Chromo2.5/STAR_genome
twopassMode Basic
outSAMtype BAM SortedByCoordinate
outSAMunmapped Within
quantMode GeneCounts
```

### Run it!

__note__ I needed to use STAR2.5 for this to work.

```{r eval=FALSE, engine='bash'}
cd spen_sim_1
STAR --parametersFiles ../STAR.params.3 --readFilesIn ./sample_01_renamed.fasta --outFileNamePrefix STAR_3 
samtools index STAR_3Aligned.sortedByCoord.out.bam
```

### Load in mapped counts
```{r }
mapped_counts.3 <- read.delim("spen_sim_1/STAR_3ReadsPerGene.out.tab",header = FALSE, stringsAsFactors = FALSE)
head(mapped_counts.3)
mapped_counts.3 <- mapped_counts.3[,1:2] # 3 and 4 are directional counts which we don't care about
colnames(mapped_counts.3) <- c("ID","count")
mapped_counts.3$ID <- sub("gene:","",mapped_counts.3$ID,fixed=TRUE)
```

### Comparison
```{r}
#check that I can use substring
unique(nchar(known_counts$id)) #all are the same length
known_counts$lyc_id <- substr(known_counts$id,18,33)
head(known_counts)
count_comparison.3 <- merge(known_counts,mapped_counts.3,by.x="lyc_id",by.y="ID",suffixes=c(".known",".mapped"))
head(count_comparison.3)
cor(count_comparison.3$count.known,count_comparison.3$count.mapped)
plot(count_comparison.3$count.known,count_comparison.3$count.mapped)
```
Not so good!

how much is chromosome zero?
```{r}
count_comparison.3$chrom <- substr(count_comparison.3$lyc_id,6,7)
head(count_comparison.3)
library(ggplot2)
qplot(x=count.known,y=count.mapped,data=count_comparison.3) + facet_wrap(~ chrom)
```
Not limited to chrom zero

Investigate...
Lets take a look at the bam file and extract reads from a few genes with 0 mapped reads.  Skip chromosome 0
```{r}
head(count_comparison.3[count_comparison.3$chrom=="01" & count_comparison.3$count.mapped==0,],n=20)
library(Rsamtools)
reads <- scanBam(file="spen_sim_1/STAR_2Aligned.sortedByCoord.out.bam",
                 param = ScanBamParam(what=scanBamWhat(),
                                      tag=c("NH","HI","AS","nM")))[[1]]
#note: DO NOT convert reads to a data frame.  Uses lots of memory.
reads <- c(reads[names(reads)!="tag"],reads$tag)
lapply(reads, head)
#pull the reads that should have mapped out into new list
zero_map_genes <- count_comparison.3$id[count_comparison.3$count.known>0 & count_comparison.3$count.mapped==0]
reads$gene <- unlist(strsplit(reads$qname,";"))[c(FALSE,TRUE,FALSE,FALSE)]
lapply(reads, head)
reads_zero <- lapply(reads,function(x) x[reads$gene %in% zero_map_genes])
lapply(reads_zero,head,20)
table(reads_zero$flag)
hist(reads_zero$NH) #number of hits
hist(reads$NH)
#so a fair number of the reads from genes with zero counts are multi-mappers.

#what about the other ones

reads_zero_1hit <- lapply(reads_zero,function(x) x[reads_zero$NH==1])
lapply(reads_zero_1hit,head,10)
```